cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/MapReduceProgramWorkflowRunner.java

```diff
@@ -61,20 +67,32 @@
 
   /**
    * Executes given {@link Program} with the given {@link ProgramOptions} and block until it completed.
-   * On completion, return the {@link RuntimeContext} of the program.
+   * On completion, return the {@link Map} of hadoop counters for the program run.
    *
    * @throws Exception if execution failed.
    */
   @Override
-  public RuntimeContext runAndWait(Program program, ProgramOptions options) throws Exception {
+  public Map<String, String> runAndWait(Program program, ProgramOptions options) throws Exception {
     ProgramController controller = programRunnerFactory.create(ProgramRunnerFactory.Type.MAPREDUCE).run(program,
                                                                                                         options);
     if (controller instanceof MapReduceProgramController) {
-      final RuntimeContext context = ((MapReduceProgramController) controller).getContext();
-      return executeProgram(controller, context);
+      MapReduceContext context = ((MapReduceProgramController) controller).getContext();
+      executeProgram(controller, context);
+      return getCounterMap(context);
     } else {
       throw new IllegalStateException("Failed to run program. The controller is not an instance of " +
                                         "MapReduceProgramController");
     }
   }
+
+  private Map<String, String> getCounterMap(MapReduceContext context) throws Exception {
+    Map<String, String> mapCounters = Maps.newHashMap();
+    Counters counters = ((Job) context.getHadoopJob()).getCounters();
+    for (CounterGroup group : counters) {
+      for (Counter counter : group) {
+        mapCounters.put(group.getName() + "." + counter.getName(), Long.toString(counter.getValue()));
```

Comment:
I mean - I don't think that resolves the problem.
With those two metrics, it'll still conflict.